pkg/controller/endpoint/endpoints_controller_cache.go

```
@@ -0,0 +1,232 @@
+// consumption.
+type endpointsControllerCache struct {
+	pods     map[ /* namespace */ string]map[ /* name */ string]*v1.Pod
+	services map[ /* namespace */ string]map[ /* name */ string]*v1.Service
+
+	// Mutex guarding this util.
+	mutex sync.Mutex
+}
+
+// Creates new instance of the endpointsControllerCache.
+func newEndpointsControllerCache() *endpointsControllerCache {
+	return &endpointsControllerCache{
+		pods:     make(map[ /* namespace */ string]map[ /* name */ string]*v1.Pod),
+		services: make(map[ /* namespace */ string]map[ /* name */ string]*v1.Service),
+	}
+}
+
+// Initialize initializes the cache with pods and services obtained from the provided listers.
+// Pods and services are listed behind mutex, which temporarily disables event listener callbacks
+// (e.g. AddPod, UpdateService, etc.). It guarantees that no information will be lost.
+// This method should be called exactly once on the startup, before the first call of the
+// GetServiceAndPods method, after the lister have synced. Please note that this initialization
+// scheme relies on the assumption that, when a handler is registered in the Informer, the current
+// state will be fully delivered to the event handler via the OnAdd callbacks.
+// TODO(mm4tt): Support initialization natively in the Informer framework and get rid of this
+// 							method. This will require a new signal (and callback in the event handler) informing
+//              that the state has been synced and fully delivered to the event handlers via the
+//              OnAdd callbacks.
+func (e *endpointsControllerCache) Initialize(serviceLister corelisters.ServiceLister, podLister corelisters.PodLister) {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+
+	services, err := serviceLister.List(labels.Everything())
+	if err != nil {
+		// Error shouldn't happened as we're getting objects from a local cache, so fail if it does.
+		klog.Fatalf("unable to list services due to %v", err)
```

I think we should return an error and handle it level above (in endpointsController) or even propagate is higher.